include/sparrow/arrow_array.hpp

```diff
@@ -0,0 +1,300 @@
+
+    arrow_array_private_data::arrow_array_private_data(
+        std::vector<buffer<std::uint8_t>> buffers,
+        std::optional<std::vector<arrow_array_shared_ptr>> children,
+        arrow_array_shared_ptr dictionary
+    )
+        : m_buffers(std::move(buffers))
+        , m_buffers_pointers(to_raw_ptr_vec<std::uint8_t>(m_buffers))
+        , m_children(std::move(children))
+        , m_children_pointers(to_raw_ptr_vec<ArrowArray>(m_children))
+        , m_dictionary(std::move(dictionary))
+    {
+    }
+
+    [[nodiscard]] inline const std::vector<buffer<std::uint8_t>>&
+    arrow_array_private_data::buffers() const noexcept
+    {
+        return m_buffers;
+    }
+
+    template <class T>
+    [[nodiscard]] const T** arrow_array_private_data::buffers_ptrs() noexcept
+    {
+        return const_cast<const T**>(reinterpret_cast<T**>(m_buffers_pointers.data()));
+    }
+
+    [[nodiscard]] inline const std::optional<std::vector<arrow_array_shared_ptr>>&
+    arrow_array_private_data::children() const noexcept
+    {
+        return m_children;
+    }
+
+    [[nodiscard]] inline ArrowArray** arrow_array_private_data::children_pointers() noexcept
+    {
+        if (m_children_pointers.empty())
+        {
+            return nullptr;
+        }
+        return m_children_pointers.data();
+    }
+
+    [[nodiscard]] inline const arrow_array_shared_ptr& arrow_array_private_data::dictionary() const noexcept
+    {
+        return m_dictionary;
+    }
+
+    [[nodiscard]] inline ArrowArray* arrow_array_private_data::dictionary_pointer() noexcept
+    {
+        return m_dictionary.get();
+    }
+
+    /**
+     * Deletes an ArrowArray.
+     *
+     * @param array The ArrowArray to delete. Should be a valid pointer to an ArrowArray. Its release callback
+     * should be set.
+     */
+    inline void delete_array(ArrowArray* array)
+    {
+        SPARROW_ASSERT_FALSE(array == nullptr)
+
+        array->buffers = nullptr;
+        array->n_buffers = 0;
+        array->length = 0;
+        array->null_count = 0;
+        array->offset = 0;
+        array->n_children = 0;
+        array->children = nullptr;
+        array->dictionary = nullptr;
+        if (array->private_data != nullptr)
+        {
+            const auto private_data = static_cast<arrow_array_private_data*>(array->private_data);
+            delete private_data;
+        }
+        array->private_data = nullptr;
+        array->release = nullptr;
+    }
+
+    inline arrow_array_unique_ptr default_arrow_array_unique_ptr()
+    {
+        auto ptr = arrow_array_unique_ptr(new ArrowArray());
+        ptr->length = 0;
+        ptr->null_count = 0;
+        ptr->offset = 0;
+        ptr->n_buffers = 0;
+        ptr->n_children = 0;
+        ptr->buffers = nullptr;
+        ptr->children = nullptr;
+        ptr->dictionary = nullptr;
+        ptr->release = nullptr;
+        ptr->private_data = nullptr;
+        return ptr;
+    }
+
+    template <class B, class C, class D>
```

Comment:
Why the need for this declaration? The definition is just after?

Reply:
Needed at some point but it's not the case anymore. Let's remove it?